src/combinatorics/additive/pluennecke_ruzsa.lean

```diff
@@ -0,0 +1,190 @@
+begin
+  rintro A' hAA',
+  obtain rfl | hA' := A'.eq_empty_or_nonempty,
+  { simp },
+  have hA₀ : (0 : ℚ) < A.card := cast_pos.2 hA.card_pos,
+  have hA₀' : (0 : ℚ) < A'.card := cast_pos.2 hA'.card_pos,
+  exact_mod_cast (div_le_div_iff (hA₀) hA₀').1 (h _ $ mem_erase_of_ne_of_mem hA'.ne_empty $
+    mem_powerset.2 $ hAA'.trans hAB),
+end
+
+/-- **Ruzsa's triangle inequality** -/
+lemma card_add_mul_le_card_add_mul_card_add (A B C : finset α) :
+  (A + C).card * B.card ≤ (A + B).card * (B + C).card :=
+begin
+  obtain rfl | hB := B.eq_empty_or_nonempty,
+  { simp },
+  have hB' : B ∈ B.powerset.erase ∅ := mem_erase_of_ne_of_mem hB.ne_empty (mem_powerset_self _),
+  obtain ⟨U, hU, hUA⟩ := exists_min_image (B.powerset.erase ∅) (λ U, (U + A).card/U.card : _ → ℚ)
+    ⟨B, hB'⟩,
+  rw [mem_erase, mem_powerset, ←nonempty_iff_ne_empty] at hU,
+  refine cast_le.1 (_ : (_ : ℚ) ≤ _),
+  push_cast,
+  refine (le_div_iff $ by exact cast_pos.2 hB.card_pos).1 _,
+  rw [mul_div_right_comm, add_comm _ B],
+  refine (cast_le.2 $ card_le_card_add_left _ hU.1).trans _,
+  refine le_trans _ (mul_le_mul (hUA _ hB') (cast_le.2 $ card_le_of_subset $
+    add_subset_add_right hU.2) (cast_nonneg _) $ div_nonneg (cast_nonneg _) $
+    cast_nonneg _),
+  rw [←mul_div_right_comm, ←add_assoc],
+  refine (le_div_iff $ by exact cast_pos.2 hU.1.card_pos).2 _,
+  exact_mod_cast pluennecke_petridis C (aux hU.1 hU.2 hUA),
+end
+
+/-- **Ruzsa's triangle inequality** -/
+lemma card_add_mul_le_card_sub_mul_card_sub (A B C : finset α) :
+  (A + C).card * B.card ≤ (A - B).card * (B - C).card :=
+begin
+  rw [sub_eq_add_neg, ←card_neg B, ←card_neg (B - C), neg_sub', sub_neg_eq_add],
+  exact card_add_mul_le_card_add_mul_card_add _ _ _,
+end
+
+/-- **Ruzsa's triangle inequality** -/
+lemma card_sub_mul_le_card_add_mul_card_sub (A B C : finset α) :
+  (A - C).card * B.card ≤ (A + B).card * (B - C).card :=
+by { rw [sub_eq_add_neg, sub_eq_add_neg], exact card_add_mul_le_card_add_mul_card_add _ _ _ }
+
+/-- **Ruzsa's triangle inequality** -/
+lemma card_sub_mul_le_card_sub_mul_card_add (A B C : finset α) :
+  (A - C).card * B.card ≤ (A - B).card * (B + C).card :=
+by { rw [←sub_neg_eq_add, sub_eq_add_neg], exact card_add_mul_le_card_sub_mul_card_sub _ _ _ }
+
+lemma card_add_nsmul_le (hAB : ∀ A' ⊆ A, (A + B).card * A'.card ≤ (A' + B).card * A.card) (n : ℕ) :
+  ((A + n • B).card : ℚ) ≤ ((A + B).card / A.card) ^ n * A.card :=
+begin
+  obtain rfl | hA := A.eq_empty_or_nonempty,
+  { simp },
+  induction n with n ih,
+  { simp },
+  rw [succ_nsmul, ←add_assoc, pow_succ, mul_assoc, ←mul_div_right_comm, le_div_iff, ←cast_mul],
+  swap, exact (cast_pos.2 hA.card_pos),
+  refine (cast_le.2 $ pluennecke_petridis _ hAB).trans _,
+  rw cast_mul,
+  exact mul_le_mul_of_nonneg_left ih (cast_nonneg _),
+end
+
+/-- The **Plünnecke-Ruzsa inequality**. -/
+lemma card_smul_sub_smul_le (hB : B.nonempty) (m n : ℕ) :
+  ((m • B - n • B).card : ℚ) ≤ ((B + B).card / B.card) ^ (m + n) * B.card :=
```

Comment:
This is a weaker version than the one on wikipedia - can we prove the more general version with A,B being different?